keps/sig-auth/NNNN-20201209-client-exec-proxy/README.md

```diff
@@ -0,0 +1,711 @@
+```go
+type ExecProxyConfig struct {
+	// The socket path where the proxy should listen.
+	UnixSocketPath string `json:"unixSocketPath"`
+}
+```
+
+Input and output from the executed binary.
+```go
+type ExecProxyRequest struct {
+	metav1.TypeMeta
+
+	// +optional
+	Config ExecProxyConfig `json:"config"`
+
+	// +optional
+	// Whether or not the command is being called from an interactive prompt.  A hardware security module might require user interaction on setup,
+	// so in that case the proxy might want to fail if this is false.
+	Interactive bool `json:"interactive"`
+
+	// +optional
+	// The destination Kubernetes cluster.
+	Cluster *Cluster `json:"cluster"`
+}
+
+type ExecProxyResponse struct {
+	metav1.TypeMeta
+}
+```
+
+Example kubeconfig:
+
+```
+- name: nic@cluster1
+  user:
+    exec:
+      apiVersion: client.authentication.k8s.io/v1alpha1
+      args:
+      - --region=us-west-2
+      command: /usr/local/bin/sigv4proxy
+      execProxy:
+        unixSocketPath: /var/run/kubelet/signer.sock
+```
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+### Graduation Criteria
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. The KEP
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha -> Beta Graduation
```

Comment:
For this to graduate to beta I would like to see an example plugin command that does TLS offload based on some local key pair (i.e. the TLS bits to the API server are being handled by the proxy and not kubectl).  This would then be used in e2e tests to prove out the functionality.